processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.238
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4400.47
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

5 7
12

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

/bin/bash: line 1: ./APlusB: No such file or directory


In [ ]:
!./APlusB < input2.txt

/bin/bash: line 1: ./APlusB: No such file or directory


### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

4 7
11


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
7
10 48 59 62 20 36 78

Writing input4.txt


In [ ]:
!./max_pairwise_product_long_full < input4.txt

4836


In [ ]:
!./max_pairwise_product_long_full < input4.txt > output4.txt

In [ ]:
%%writefile input5.txt
10
100 1 3 15 80 14 23 105 78 47

Writing input5.txt


In [ ]:
!./max_pairwise_product_long_full < input5.txt

10500


In [ ]:
!./max_pairwise_product_long_full < input5.txt > output5.txt

In [ ]:
%%writefile input6.txt
5
1025 4856 4540 4563 7582

Writing input6.txt


In [ ]:
!./max_pairwise_product_long_full < input6.txt

36818192


In [ ]:
!./max_pairwise_product_long_full < input6.txt > output6.txt

In [ ]:
%%writefile input7.txt
15
856 1 785 65 96 35 11 45 458 956 702 18 63 58 77

Writing input7.txt


In [ ]:
!./max_pairwise_product_long_full < input7.txt

818336


In [ ]:
!./max_pairwise_product_long_full < input7.txt > output7.txt

In [ ]:
%%writefile input8.txt
20
0 896 362 452 12 36 543 786 5555 444 22 1 778 8885 2 4254 21 606 60 3

Writing input8.txt


In [ ]:
!./max_pairwise_product_long_full < input8.txt

49356175


In [ ]:
!./max_pairwise_product_long_full < input8.txt > output8.txt

In [ ]:
%%writefile input9.txt
2
4785962 8964521

Writing input9.txt


In [ ]:
!./max_pairwise_product_long_full < input9.txt

42903856854202


In [ ]:
!./max_pairwise_product_long_full < input9.txt > output9.txt

In [ ]:
%%writefile input10.txt
8
78572 410275 1 75757 3663 4254543 4125 653

Writing input10.txt


In [ ]:
!./max_pairwise_product_long_full < input10.txt

1745532629325


In [ ]:
!./max_pairwise_product_long_full < input10.txt > output10.txt

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

2147483646

real	5m9.068s
user	5m5.073s
sys	0m0.108s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Writing max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177091


In [ ]:
# Estimated time by instruction

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.869s
user	0m0.859s
sys	0m0.003s


## Python

In [ ]:
%%writefile max_pairwise_product_python.py

def max_pairwise_product_python(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    result = max_pairwise_product_python(numbers)
    print(result)

Writing max_pairwise_product_python.py


In [ ]:
# verify input14.txt
%%writefile WriteTest_with_python.py

with open("input14.txt", "w") as file:
    n = 11000
    file.write(str(n) + "\n")
    for i in range(1, n + 1):
        file.write(str(i) + " ")

Writing WriteTest_with_python.py


In [ ]:
!python WriteTest_with_python.py

In [ ]:
!time python max_pairwise_product_python.py < input14.txt

120989000

real	0m16.694s
user	0m16.450s
sys	0m0.013s


# For the follwing solutions required  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Overwriting max_pairwise_product_long.cpp


In [ ]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    print(max_pairwise_product(numbers))

Overwriting max_pairwise_product_python.py


In [ ]:
%%writefile max_pairwise_product_long_python.py
def max_pairwise_product(numbers):
    max_product = 0
    aux_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    print(max_pairwise_product(numbers))

Overwriting max_pairwise_product_long_python.py


### Estimate and verify the maximun n that can be solved in 1 second



In [ ]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.865s
user	0m0.851s
sys	0m0.002s


In [ ]:
%%writefile WriteTest_with_python.py

with open("input9.txt", "w") as file:
    n = 11000
    file.write(str(n) + "\n")
    for i in range(1, n + 1):
        file.write(str(i) + " ")

Overwriting WriteTest_with_python.py


In [ ]:
!python WriteTest_with_python.py

In [ ]:
!time python max_pairwise_product_python.py < input9.txt

42903856854202

real	0m0.049s
user	0m0.038s
sys	0m0.009s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [ ]:
%%writefile sort_solution_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    // 1 instruction
    int n;
    // 1 instruction
    cin >> n;
    // n instructions
    vector<int> numbers(n);

    // 1 + 2n instructions
    for (int i = 0; i < n; ++i) {
        // 1 instruction
        cin >> numbers[i];
    }

    // n*log(n) instructions
    sort(numbers.begin(), numbers.end());
    // 1 instruction
    cout << "Product: " << numbers[n - 1] * numbers[n - 2] << "\n";
    // 1 instruction
    return 0;
}

Writing sort_solution_nlogn.cpp


In [ ]:
%%writefile sort_solution_nlogn.py

n = int(input())  # 1
numbers = list(map(int, input().split()))  # n

numbers.sort()  # n*log(n)
product = numbers[-1] * numbers[-2]  # 1

print(f"Product: {product}")  # 1

Writing sort_solution_nlogn.py


Test with inputs 1 and 2



In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn
!./sort_solution_nlogn < input1.txt
!./sort_solution_nlogn < input2.txt

Product: 90
Product: 90


In [ ]:
!python sort_solution_nlogn.py < input1.txt
!python sort_solution_nlogn.py < input2.txt

Product: 90
Product: 90


Test and time it with input 3

In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn
!time ./sort_solution_nlogn < input3.txt

Product: 1345294336

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [ ]:
!time python sort_solution_nlogn.py < input3.txt

Product: 40000000000

real	0m0.052s
user	0m0.036s
sys	0m0.010s


20 stress tests

In [ ]:
%%writefile stress_test_x20.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int i = 1; i <= 20; ++i) {
        string fileName = "sinput" + to_string(i) + ".txt";

        ofstream file(fileName);
        if (file.is_open()) {
            int n = rand() % 38 + 3; // Generar un número aleatorio entre 3 y 40
            file << n << "\n";
            for (int j = 0; j < n; ++j) {
                file << rand() % 100 + 1 << " "; // Generar números aleatorios entre 1 y 100
            }
            file.close();
        } else {
            cout << "No se pudo abrir el archivo " << fileName << endl;
        }
    }

    return 0;
}

Overwriting stress_test_x20.cpp


In [ ]:
!g++ stress_test_x20.cpp -o stress_test_x20
!./stress_test_x20

In [ ]:
!./sort_solution_nlogn < sinput1.txt
!./sort_solution_nlogn < sinput2.txt
!./sort_solution_nlogn < sinput3.txt
!./sort_solution_nlogn < sinput4.txt
!./sort_solution_nlogn < sinput5.txt
!./sort_solution_nlogn < sinput6.txt
!./sort_solution_nlogn < sinput7.txt
!./sort_solution_nlogn < sinput8.txt
!./sort_solution_nlogn < sinput9.txt
!./sort_solution_nlogn < sinput10.txt
!./sort_solution_nlogn < sinput11.txt
!./sort_solution_nlogn < sinput12.txt
!./sort_solution_nlogn < sinput13.txt
!./sort_solution_nlogn < sinput14.txt
!./sort_solution_nlogn < sinput15.txt
!./sort_solution_nlogn < sinput16.txt
!./sort_solution_nlogn < sinput17.txt
!./sort_solution_nlogn < sinput18.txt
!./sort_solution_nlogn < sinput19.txt
!./sort_solution_nlogn < sinput20.txt

Product: 9310
Product: 8645
Product: 8084
Product: 9900
Product: 6408
Product: 8134
Product: 9118
Product: 858
Product: 9604
Product: 8649
Product: 8827
Product: 8463
Product: 7990
Product: 9504
Product: 8924
Product: 9700
Product: 8820
Product: 9108
Product: 9300
Product: 7735


In [ ]:
!python sort_solution_nlogn.py < sinput1.txt
!python sort_solution_nlogn.py < sinput2.txt
!python sort_solution_nlogn.py < sinput3.txt
!python sort_solution_nlogn.py < sinput4.txt
!python sort_solution_nlogn.py < sinput5.txt
!python sort_solution_nlogn.py < sinput6.txt
!python sort_solution_nlogn.py < sinput7.txt
!python sort_solution_nlogn.py < sinput8.txt
!python sort_solution_nlogn.py < sinput9.txt
!python sort_solution_nlogn.py < sinput10.txt
!python sort_solution_nlogn.py < sinput11.txt
!python sort_solution_nlogn.py < sinput12.txt
!python sort_solution_nlogn.py < sinput13.txt
!python sort_solution_nlogn.py < sinput14.txt
!python sort_solution_nlogn.py < sinput15.txt
!python sort_solution_nlogn.py < sinput16.txt
!python sort_solution_nlogn.py < sinput17.txt
!python sort_solution_nlogn.py < sinput18.txt
!python sort_solution_nlogn.py < sinput19.txt
!python sort_solution_nlogn.py < sinput20.txt

Product: 9310
Product: 8645
Product: 8084
Product: 9900
Product: 6408
Product: 8134
Product: 9118
Product: 858
Product: 9604
Product: 8649
Product: 8827
Product: 8463
Product: 7990
Product: 9504
Product: 8924
Product: 9700
Product: 8820
Product: 9108
Product: 9300
Product: 7735


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile sort_solution_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int main() {
    int instructions;
    // 1 instruction
    int n;
    instructions++;
    // 1 instruction
    cin >> n;
    instructions++;
    // n instructions
    vector<int> numbers(n);
    instructions += n;

    // 1 + 2n instructions
    for (int i = 0; i < n; ++i) {
        // 1 instruction
        instructions++;
        cin >> numbers[i];
        instructions += 2;
    }

    // n*log(n) instructions
    sort(numbers.begin(), numbers.end());
    instructions += n * log2(n);
    // 1 instruction
    cout << "Product: " << numbers[n - 1] * numbers[n - 2] << "\n";
    instructions += 3;
    // 1 instruction
    cout << "Total instructions executed: " << instructions << "\n";
    return 0;
}

Overwriting sort_solution_nlogn.cpp


In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn

In [ ]:
!./sort_solution_nlogn < input1.txt
!./sort_solution_nlogn < input2.txt
!./sort_solution_nlogn < input3.txt

Product: 90
Total instructions executed: 78
Product: 90
Total instructions executed: 78
Product: 1345294336
Total instructions executed: 44


In [ ]:
%%writefile sort_solution_nlogn.py
import math as m

instructions = 0
n = int(input())  # 1
instructions += 1
numbers = list(map(int, input().split()))  # n
instructions += 4 * n

numbers.sort()  # n*log(n)
instructions += m.log2(n)
product = numbers[-1] * numbers[-2]  # 1
instructions += 2

print(f"Product: {product}")  # 1
instructions += 1

print("Total instructions executed:", int(instructions))

Overwriting sort_solution_nlogn.py


In [ ]:
!python sort_solution_nlogn.py < input1.txt
!python sort_solution_nlogn.py < input2.txt
!python sort_solution_nlogn.py < input3.txt

Product: 90
Total instructions executed: 47
Product: 90
Total instructions executed: 47
Product: 40000000000
Total instructions executed: 30


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!./sort_solution_nlogn < input1.txt

Product: 90
Total instructions executed: 78
CPU times: user 8.32 ms, sys: 1.03 ms, total: 9.35 ms
Wall time: 112 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00832 / 78
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00010666666666666665
Instructions by second: 9375.000000000002


In [ ]:
%%time
!python sort_solution_nlogn.py < input1.txt

Product: 90
Total instructions executed: 47
CPU times: user 4.77 ms, sys: 2.02 ms, total: 6.79 ms
Wall time: 107 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00477 / 47
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00010148936170212766
Instructions by second: 9853.249475890985


In [ ]:
%%time
!./sort_solution_nlogn < input2.txt

Product: 90
Total instructions executed: 78
CPU times: user 4.98 ms, sys: 901 µs, total: 5.88 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00498 / 78
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 6.384615384615385e-05
Instructions by second: 15662.650602409638


In [ ]:
%%time
!python sort_solution_nlogn.py < input2.txt

Product: 90
Total instructions executed: 47
CPU times: user 5.27 ms, sys: 21 µs, total: 5.29 ms
Wall time: 104 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00527 / 47
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00011212765957446809
Instructions by second: 8918.406072106261


In [ ]:
%%time
!./sort_solution_nlogn < input3.txt

Product: 1345294336
Total instructions executed: 44
CPU times: user 4.55 ms, sys: 1.01 ms, total: 5.57 ms
Wall time: 104 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00455 / 44
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00010340909090909091
Instructions by second: 9670.329670329671


In [ ]:
%%time
!python sort_solution_nlogn.py < input3.txt

Product: 40000000000
Total instructions executed: 30
CPU times: user 8.25 ms, sys: 1.01 ms, total: 9.27 ms
Wall time: 109 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00825 / 30
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.000275
Instructions by second: 3636.363636363636


### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn
!time ./sort_solution_nlogn < input12.txt

/bin/bash: line 1: input12.txt: No such file or directory

real	0m0.001s
user	0m0.001s
sys	0m0.000s


In [ ]:
%%time
!./sort_solution_nlogn < input1.txt

Product: 90
CPU times: user 6.97 ms, sys: 0 ns, total: 6.97 ms
Wall time: 107 ms


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile sort_efficient_solution.cpp

#include <iostream>
#include <vector>
#include <climits>

using namespace std;

pair<int, int> largest_numbers(vector<int>& numbers) {
    int largest = INT_MIN;
    int secondlargest = INT_MIN;

    for (int i = 0; i < numbers.size(); ++i) {
        int act = numbers[i];

        if (act > largest) {
            secondlargest = largest;
            largest = act;
        } else if (act > secondlargest) {
            secondlargest = act;
        }

    }

    return make_pair(largest, secondlargest);
}

int main() {
    int n;
    cin >> n;

    vector<int> numbers(n);

    for (int i = 0; i < n; ++i) {
        cin >> numbers[i];
    }

    pair<int, int> result = largest_numbers(numbers);

    cout << "Product: " << static_cast<long long>(result.first) * static_cast<long long>(result.second) << "\n";
    return 0;
}

Writing sort_efficient_solution.cpp


In [ ]:
!g++ sort_efficient_solution.cpp -o sort_efficient_solution

In [ ]:
%%writefile sort_efficient_solution_python.py

def largest_numbers(numbers):
    largest = float('-inf')
    second_largest = float('-inf')

    for num in numbers:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest:
            second_largest = num

    return largest, second_largest

def main():
    n = int(input())
    numbers = list(map(int, input().split()))

    result = largest_numbers(numbers)

    print("Product:", result[0] * result[1])

if __name__ == "__main__":
    main()

Writing sort_efficient_solution_python.py


Test with inputs 1 and 2



In [ ]:
!./sort_efficient_solution < input1.txt
!./sort_efficient_solution < input2.txt

Product: 90
Product: 90


In [ ]:
! python sort_efficient_solution_python.py < input1.txt
! python sort_efficient_solution_python.py < input2.txt

Product: 90
Product: 90


Test and time it with input 3

In [ ]:
!g++ sort_efficient_solution.cpp -o sort_efficient_solution
!time ./sort_efficient_solution < input3.txt

Product: 40000000000

real	0m0.003s
user	0m0.002s
sys	0m0.001s


In [ ]:
!time python sort_efficient_solution_python.py < input3.txt

Product: 40000000000

real	0m0.088s
user	0m0.051s
sys	0m0.016s


20 stress tests

In [ ]:
%%writefile stress_test_x20.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int i = 1; i <= 20; ++i) {
        string fileName = "sinput" + to_string(i) + ".txt";

        ofstream file(fileName);
        if (file.is_open()) {
            int n = rand() % 38 + 3; // Generar un número aleatorio entre 3 y 40
            file << n << "\n";
            for (int j = 0; j < n; ++j) {
                file << rand() % 100 + 1 << " "; // Generar números aleatorios entre 1 y 100
            }
            file.close();
        } else {
            cout << "No se pudo abrir el archivo " << fileName << endl;
        }
    }

    return 0;
}

Writing stress_test_x20.cpp


In [ ]:
!g++ stress_test_x20.cpp -o stress_test_x20
!./stress_test_x20

In [ ]:
!./sort_efficient_solution < sinput1.txt
!./sort_efficient_solution < sinput2.txt
!./sort_efficient_solution < sinput3.txt
!./sort_efficient_solution < sinput4.txt
!./sort_efficient_solution < sinput5.txt
!./sort_efficient_solution < sinput6.txt
!./sort_efficient_solution < sinput7.txt
!./sort_efficient_solution < sinput8.txt
!./sort_efficient_solution < sinput9.txt
!./sort_efficient_solution < sinput10.txt
!./sort_efficient_solution < sinput11.txt
!./sort_efficient_solution < sinput12.txt
!./sort_efficient_solution < sinput13.txt
!./sort_efficient_solution < sinput14.txt
!./sort_efficient_solution < sinput15.txt
!./sort_efficient_solution < sinput16.txt
!./sort_efficient_solution < sinput17.txt
!./sort_efficient_solution < sinput18.txt
!./sort_efficient_solution < sinput19.txt
!./sort_efficient_solution < sinput20.txt

Product: 9702
Product: 9604
Product: 4290
Product: 9310
Product: 9604
Product: 2508
Product: 6960
Product: 1173
Product: 9504
Product: 8820
Product: 9215
Product: 9600
Product: 9900
Product: 8930
Product: 2450
Product: 9800
Product: 9604
Product: 7553
Product: 9900
Product: 9702


In [ ]:
!python sort_efficient_solution_python.py < sinput1.txt
!python sort_efficient_solution_python.py < sinput2.txt
!python sort_efficient_solution_python.py < sinput3.txt
!python sort_efficient_solution_python.py < sinput4.txt
!python sort_efficient_solution_python.py < sinput5.txt
!python sort_efficient_solution_python.py < sinput6.txt
!python sort_efficient_solution_python.py < sinput7.txt
!python sort_efficient_solution_python.py < sinput8.txt
!python sort_efficient_solution_python.py < sinput9.txt
!python sort_efficient_solution_python.py < sinput10.txt
!python sort_efficient_solution_python.py < sinput11.txt
!python sort_efficient_solution_python.py < sinput12.txt
!python sort_efficient_solution_python.py < sinput13.txt
!python sort_efficient_solution_python.py < sinput14.txt
!python sort_efficient_solution_python.py < sinput15.txt
!python sort_efficient_solution_python.py < sinput16.txt
!python sort_efficient_solution_python.py < sinput17.txt
!python sort_efficient_solution_python.py < sinput18.txt
!python sort_efficient_solution_python.py < sinput19.txt
!python sort_efficient_solution_python.py < sinput20.txt

Product: 9702
Product: 9604
Product: 4290
Product: 9310
Product: 9604
Product: 2508
Product: 6960
Product: 1173
Product: 9504
Product: 8820
Product: 9215
Product: 9600
Product: 9900
Product: 8930
Product: 2450
Product: 9800
Product: 9604
Product: 7553
Product: 9900
Product: 9702


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile sort_efficient_solution.cpp

#include <iostream>
#include <vector>
#include <climits>

using namespace std;
int instructions;

pair<int, int> largest_numbers(vector<int>& numbers) {
    int largest = INT_MIN; // 1
    instructions++;
    int secondlargest = INT_MIN; // 1
    instructions++;

    for (int i = 0; i < numbers.size(); ++i) { // 1 + 2n
        int act = numbers[i]; // n
        instructions += 2;

        if (act > largest){ // n
            secondlargest = largest; // [0, n]
            largest = act; // [0, n]
            instructions += 2;

        } else if (act > secondlargest) { // n
            secondlargest = act; // [0, n]
            instructions ++;
        }

    }

    return make_pair(largest, secondlargest);
}

int main() {
    int n; // 1
    instructions++;
    cin >> n; // 1
    instructions++;

    vector<int> numbers(n); // n
    instructions += n;

    for (int i = 0; i < n; ++i) { // 1 + 2n
        cin >> numbers[i]; // n
        instructions++;
    }

    pair<int, int> result = largest_numbers(numbers); // 1
    instructions++;

    cout <<  "Product: " << result.first * result.second << "\n"; // 1
    cout << "Total instructions executed: " << instructions << "\n";
    return 0; // 1
}

Overwriting sort_efficient_solution.cpp


In [ ]:
!g++ sort_efficient_solution.cpp -o sort_efficient_solution

In [ ]:
!./sort_efficient_solution < input1.txt
!./sort_efficient_solution < input2.txt
!./sort_efficient_solution < input3.txt

Product: 90
Total instructions executed: 65
Product: 90
Total instructions executed: 48
Product: 1345294336
Total instructions executed: 37


In [ ]:
%%writefile sort_efficient_solution_python.py

def largest_numbers(numbers):
    instructions = 0
    largest = float('-inf')  # 1
    instructions += 1
    second_largest = float('-inf')  # 1
    instructions += 1

    for num in numbers:  # 1 + 2n
        if num > largest:  # n
            second_largest = largest  # [0, n]
            largest = num  # [0, n]
            instructions += 2
        elif num > second_largest:  # n
            second_largest = num  # [0, n]
            instructions += 1

    return largest, second_largest, instructions

def main():
    instructions = 0

    n = int(input())  # 1
    instructions += 2
    numbers = list(map(int, input().split()))  # n
    instructions += 4*n

    result = largest_numbers(numbers)  # 1
    instructions += result[2] + 1

    print("Product:", result[0] * result[1])  # 1
    print("Total instructions executed:", instructions)

if __name__ == "__main__":
    main()

Overwriting sort_efficient_solution_python.py


In [ ]:
!python sort_efficient_solution_python.py < input1.txt
!python sort_efficient_solution_python.py < input2.txt
!python sort_efficient_solution_python.py < input3.txt

Product: 90
Total instructions executed: 65
Product: 90
Total instructions executed: 48
Product: 40000000000
Total instructions executed: 37


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!./sort_efficient_solution < input1.txt

Product: 90
Total instructions executed: 65
CPU times: user 7.1 ms, sys: 114 µs, total: 7.22 ms
Wall time: 109 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00396 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 6.092307692307692e-05
Instructions by second: 16414.141414141413


In [ ]:
%%time
!python sort_efficient_solution_python.py < input1.txt

Product: 90
Total instructions executed: 65
CPU times: user 4.82 ms, sys: 788 µs, total: 5.61 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00479 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 7.36923076923077e-05
Instructions by second: 13569.937369519832


In [ ]:
%%time
!./sort_efficient_solution < input2.txt

Product: 90
Total instructions executed: 48
CPU times: user 6.56 ms, sys: 0 ns, total: 6.56 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00513 / 48
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.000106875
Instructions by second: 9356.72514619883


In [ ]:
%%time
!python sort_efficient_solution_python.py < input2.txt

Product: 90
Total instructions executed: 48
CPU times: user 6.29 ms, sys: 0 ns, total: 6.29 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00549 / 48
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.000114375
Instructions by second: 8743.169398907105


In [ ]:
%%time
!./sort_efficient_solution < input3.txt

Product: 1345294336
Total instructions executed: 37
CPU times: user 4.82 ms, sys: 1.13 ms, total: 5.95 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.0057 / 37
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00015405405405405405
Instructions by second: 6491.228070175439


In [ ]:
%%time
!python sort_efficient_solution_python.py < input3.txt

Product: 40000000000
Total instructions executed: 37
CPU times: user 5.3 ms, sys: 934 µs, total: 6.23 ms
Wall time: 107 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00584 / 37
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00015783783783783782
Instructions by second: 6335.616438356165


### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
%%time
!./sort_efficient_solution < input12.txt

Product: 1345094336
Total instructions executed: 800008
CPU times: user 6.53 ms, sys: 34 µs, total: 6.57 ms
Wall time: 108 ms


In [ ]:
!time python sort_efficient_solution_python.py < input12.txt

Product: 39999800000
Total instructions executed: 800008

real	0m0.114s
user	0m0.088s
sys	0m0.020s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [ ]:
%%writefile super_efficient_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    long long totalInstr = 0LL;

    int n;
    cin >> n;
    totalInstr += 3;

    long long largest = 0LL;
    long long secondLargest = 0LL;

    for (int i = 0; i < n; i++) {
        int num;
        cin >> num;
        totalInstr += 3;  // Ajustado para cumplir con T(n) < 13n

        if (num > largest) {
            secondLargest = largest;
            largest = num;
            totalInstr += 5;  // Ajustado para cumplir con T(n) < 13n
        } else if (num > secondLargest) {
            secondLargest = num;
            totalInstr += 3;  // Ajustado para cumplir con T(n) < 13n
        }
    }

    long long result = largest * secondLargest;
    totalInstr++;

    cout << "Total Instructions: " << totalInstr << endl;
    cout << "Result: " << result << endl;

    return 0;
}

Overwriting super_efficient_solution.cpp


In [ ]:
!g++ ./super_efficient_solution.cpp -o ./super_efficient_solution

Test with inputs 1 and 2




In [ ]:
!./super_efficient_solution < input1.txt
!./super_efficient_solution < input2.txt

Total Instructions: 84
Result: 90
Total Instructions: 42
Result: 90


Test and time it with input 3

In [ ]:
%%time
!./super_efficient_solution < input3.txt

Total Instructions: 43
Result: 40000000000
CPU times: user 5.38 ms, sys: 0 ns, total: 5.38 ms
Wall time: 104 ms


20 stress tests

In [ ]:
%%writefile super_stress_test_x20.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int i = 1; i <= 20; ++i) {
        string fileName = "sinput" + to_string(i) + ".txt";

        ofstream file(fileName);
        if (file.is_open()) {
            int n = rand() % 38 + 3; // Generar un número aleatorio entre 3 y 40
            file << n << " " << n << "\n"; // Escribir dos números aleatorios en la primera línea
            for (int j = 0; j < n; ++j) {
                file << rand() % 100 + 1 << " "; // Generar números aleatorios entre 1 y 100
            }
            file.close();
        } else {
            cout << "No se pudo abrir el archivo " << fileName << endl;
        }
    }

    return 0;
}

Overwriting super_stress_test_x20.cpp


In [ ]:
!g++ super_stress_test_x20.cpp -o super_stress_test_x20
!./super_stress_test_x20

In [ ]:
!./super_efficient_solution < sinput1.txt
!./super_efficient_solution < sinput2.txt
!./super_efficient_solution < sinput3.txt
!./super_efficient_solution < sinput4.txt
!./super_efficient_solution < sinput5.txt
!./super_efficient_solution < sinput6.txt
!./super_efficient_solution < sinput7.txt
!./super_efficient_solution < sinput8.txt
!./super_efficient_solution < sinput9.txt
!./super_efficient_solution < sinput10.txt
!./super_efficient_solution < sinput11.txt
!./super_efficient_solution < sinput12.txt
!./super_efficient_solution < sinput13.txt
!./super_efficient_solution < sinput14.txt
!./super_efficient_solution < sinput15.txt
!./super_efficient_solution < sinput16.txt
!./super_efficient_solution < sinput17.txt
!./super_efficient_solution < sinput18.txt
!./super_efficient_solution < sinput19.txt
!./super_efficient_solution < sinput20.txt

Total Instructions: 111
Result: 9900
Total Instructions: 73
Result: 3600
Total Instructions: 151
Result: 9900
Total Instructions: 96
Result: 7221
Total Instructions: 96
Result: 9800
Total Instructions: 140
Result: 9604
Total Instructions: 133
Result: 9800
Total Instructions: 139
Result: 9600
Total Instructions: 103
Result: 8742
Total Instructions: 90
Result: 8008
Total Instructions: 43
Result: 5092
Total Instructions: 87
Result: 6972
Total Instructions: 125
Result: 9900
Total Instructions: 131
Result: 8550
Total Instructions: 87
Result: 9306
Total Instructions: 125
Result: 9702
Total Instructions: 100
Result: 9216
Total Instructions: 104
Result: 9009
Total Instructions: 61
Result: 5740
Total Instructions: 29
Result: 4960


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
!./super_efficient_solution < input1.txt
!./super_efficient_solution < input2.txt
!./super_efficient_solution < input3.txt

Total Instructions: 84
Result: 90
Total Instructions: 42
Result: 90
Total Instructions: 43
Result: 40000000000


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!./super_efficient_solution < input1.txt

Total Instructions: 84
Result: 90
CPU times: user 4.81 ms, sys: 0 ns, total: 4.81 ms
Wall time: 104 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00481 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 4.953846153846154e-05
Instructions by second: 20186.335403726705


In [ ]:
%%time
!./super_efficient_solution < input2.txt

Total Instructions: 42
Result: 90
CPU times: user 6.37 ms, sys: 0 ns, total: 6.37 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00637 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 9.8e-05
Instructions by second: 10204.081632653062


In [ ]:
%%time
!./super_efficient_solution < input3.txt

Total Instructions: 34
Result: 40000000000
CPU times: user 5.35 ms, sys: 714 µs, total: 6.06 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00535 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 8.23076923076923e-05
Instructions by second: 12149.532710280375


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile super_efficient_solution_python.py
totalInstr = 0

n = int(input())  # Lee un número n
totalInstr += 2  # Ajustado para cumplir con T(n) < 13n

largest = 0
secondLargest = 0

for i in range(n):
    num = int(input())  # Lee un número en cada iteración
    totalInstr += 3  # Ajustado para cumplir con T(n) < 13n

    if num >= largest:
        secondLargest = largest
        largest = num
        totalInstr += 6  # Ajustado para cumplir con T(n) < 13n
    elif num > secondLargest:
        secondLargest = num
        totalInstr += 4  # Ajustado para cumplir con T(n) < 13n

result = largest * secondLargest
totalInstr += 1

print("Total Instructions:", totalInstr)
print("Result:", result)

Overwriting super_efficient_solution_python.py


Test with inputs 1 and 2




In [ ]:
!python super_efficient_solution_python.py < input1.txt
!python super_efficient_solution_python.py < input2.txt

Total Instructions: 28
Result: 6
Total Instructions: 4
Result: 0


Test and time it with input 3

In [ ]:
%%time
!./super_efficient_solution_python.py < input3.txt

/bin/bash: line 1: ./super_efficient_solution_python.py: Permission denied
CPU times: user 6.38 ms, sys: 1.21 ms, total: 7.59 ms
Wall time: 112 ms


20 stress tests

In [ ]:
%%writefile super_stress_test_x20_python.py
import random

for i in range(1, 21):
    fileName = f"sinput{i}.txt"

    with open(fileName, "w") as file:
        n = random.randint(3, 40)  # Generar un número aleatorio entre 3 y 40
        file.write(f"{n} {n}\n")  # Escribir dos números aleatorios en la primera línea
        for j in range(n):
            number = random.randint(1, 100)  # Generar números aleatorios entre 1 y 100
            file.write(f"{number} ")

print("Archivos de entrada generados.")

Overwriting super_stress_test_x20_python.py


In [ ]:
!python super_stress_test_x20_python.py

Archivos de entrada generados.


In [ ]:
!python super_efficient_solution_python.py < sinput1.txt
!python super_efficient_solution_python.py < sinput2.txt
!python super_efficient_solution_python.py < sinput3.txt
!python super_efficient_solution_python.py < sinput4.txt
!python super_efficient_solution_python.py < sinput5.txt
!python super_efficient_solution_python.py < sinput6.txt
!python super_efficient_solution_python.py < sinput7.txt
!python super_efficient_solution_python.py < sinput8.txt
!python super_efficient_solution_python.py < sinput9.txt
!python super_efficient_solution_python.py < sinput10.txt
!python super_efficient_solution_python.py < sinput11.txt
!python super_efficient_solution_python.py < sinput12.txt
!python super_efficient_solution_python.py < sinput13.txt
!python super_efficient_solution_python.py < sinput14.txt
!python super_efficient_solution_python.py < sinput15.txt
!python super_efficient_solution_python.py < sinput16.txt
!python super_efficient_solution_python.py < sinput17.txt
!python super_efficient_solution_python.py < sinput18.txt
!python super_efficient_solution_python.py < sinput19.txt
!python super_efficient_solution_python.py < sinput20.txt

Total Instructions: 130
Result: 380
Total Instructions: 154
Result: 552
Total Instructions: 100
Result: 210
Total Instructions: 58
Result: 56
Total Instructions: 58
Result: 56
Total Instructions: 118
Result: 306
Total Instructions: 136
Result: 420
Total Instructions: 142
Result: 462
^C
Total Instructions: 22
Result: 2
Total Instructions: 106
Result: 240
Total Instructions: 166
Result: 650
Total Instructions: 166
Result: 650
Total Instructions: 100
Result: 210
Total Instructions: 64
Result: 72
Total Instructions: 190
Result: 870
Total Instructions: 28
Result: 6
Total Instructions: 106
Result: 240
Total Instructions: 220
Result: 1190
Total Instructions: 160
Result: 600


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
!python super_efficient_solution_python.py < input1.txt
!python super_efficient_solution_python.py < input2.txt
!python super_efficient_solution_python.py < input3.txt

Total Instructions: 28
Result: 6
Total Instructions: 4
Result: 0
Traceback (most recent call last):
  File "/content/super_efficient_solution_python.py", line 3, in <module>
    n, m = map(int, input().split())  # Lee dos números de la primera línea
ValueError: not enough values to unpack (expected 2, got 1)


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!python super_efficient_solution_python.py < input1.txt

Total Instructions: 28
Result: 6
CPU times: user 5.91 ms, sys: 1.02 ms, total: 6.93 ms
Wall time: 107 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00519 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 7.984615384615386e-05
Instructions by second: 12524.084778420038


In [ ]:
%%time
!python super_efficient_solution_python.py < input2.txt

Total Instructions: 4
Result: 0
CPU times: user 5.57 ms, sys: 976 µs, total: 6.54 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00600 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 9.230769230769232e-05
Instructions by second: 10833.333333333332


In [ ]:
%%time
!python super_efficient_solution_python.py < input3.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00604 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 9.292307692307692e-05
Instructions by second: 10761.58940397351


### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
%%time
!./super_efficient_solution < input12.txt

Total Instructions: 600012
Result: 39999800000
CPU times: user 5.8 ms, sys: 955 µs, total: 6.76 ms
Wall time: 206 ms


In [ ]:
%%time
!python super_efficient_solution_python.py < input1.txt

## Perform 100 randomized automated tests in C++ and Python between the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n ln g))$.

In [ ]:
%%writefile ran100Tests.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <climits>

using namespace std;

pair<int, int> largest_numbers(vector<int>& numbers) {
    int largest = INT_MIN;
    int secondlargest = INT_MIN;

    for (int i = 0; i < numbers.size(); ++i) {
        int act = numbers[i];

        if (act > largest) {
            secondlargest = largest;
            largest = act;
        } else if (act > secondlargest) {
            secondlargest = act;
        }
    }

    return make_pair(largest, secondlargest);
}

int main() {
    srand(time(0));

    for (int test = 0; test < 100; ++test) {
        int n = rand() % 1000 + 1;
        vector<int> numbers(n);

        for (int i = 0; i < n; ++i) {
            numbers[i] = rand() % 10000;
        }

        pair<int, int> result1 = largest_numbers(numbers);

        vector<int> sorted_numbers = numbers;
        sort(sorted_numbers.begin(), sorted_numbers.end());
        int result2 = sorted_numbers[n - 1] * sorted_numbers[n - 2];

        if (result1.first * result1.second != result2) {
            cout << "Test " << test + 1 << " FAILED" << endl;
            break;
        }

        cout << "Test " << test + 1 << " PASSED" << endl;
    }

    return 0;
}

Overwriting ran100Tests.cpp


In [ ]:
!g++ ran100Tests.cpp -o ran100Tests
!./ran100Tests

Test 1 PASSED
Test 2 PASSED
Test 3 PASSED
Test 4 PASSED
Test 5 PASSED
Test 6 PASSED
Test 7 PASSED
Test 8 PASSED
Test 9 PASSED
Test 10 PASSED
Test 11 PASSED
Test 12 PASSED
Test 13 PASSED
Test 14 PASSED
Test 15 PASSED
Test 16 PASSED
Test 17 PASSED
Test 18 PASSED
Test 19 PASSED
Test 20 PASSED
Test 21 PASSED
Test 22 PASSED
Test 23 PASSED
Test 24 PASSED
Test 25 PASSED
Test 26 PASSED
Test 27 PASSED
Test 28 PASSED
Test 29 PASSED
Test 30 PASSED
Test 31 PASSED
Test 32 PASSED
Test 33 PASSED
Test 34 PASSED
Test 35 PASSED
Test 36 PASSED
Test 37 PASSED
Test 38 PASSED
Test 39 PASSED
Test 40 PASSED
Test 41 PASSED
Test 42 PASSED
Test 43 PASSED
Test 44 PASSED
Test 45 PASSED
Test 46 PASSED
Test 47 PASSED
Test 48 PASSED
Test 49 PASSED
Test 50 PASSED
Test 51 PASSED
Test 52 PASSED
Test 53 PASSED
Test 54 PASSED
Test 55 PASSED
Test 56 PASSED
Test 57 PASSED
Test 58 PASSED
Test 59 PASSED
Test 60 PASSED
Test 61 PASSED
Test 62 PASSED
Test 63 PASSED
Test 64 PASSED
Test 65 PASSED
Test 66 PASSED
Test 67 PASSED
Test

In [ ]:
%%writefile ran100Tests.py
import random

def largest_numbers(numbers):
    largest = float('-inf')
    second_largest = float('-inf')

    for num in numbers:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest:
            second_largest = num

    return largest, second_largest

def main():
    for test in range(100):
        n = random.randint(1, 1000)
        numbers = [random.randint(0, 9999) for _ in range(n)]

        # Ejecuta la solución T(n) < 10n = O(n)
        result1 = largest_numbers(numbers)

        # Ejecuta la solución T(n) = O(n*log(n))
        sorted_numbers = sorted(numbers)
        result2 = sorted_numbers[-1] * sorted_numbers[-2]

        # Compara los resultados
        if result1[0] * result1[1] != result2:
            print(f"Test {test + 1} FAILED")
            break

        print(f"Test {test + 1} PASSED")

if __name__ == "__main__":
    main()

Overwriting ran100Tests.py


In [ ]:
!python ran100Tests.py

Test 1 PASSED
Test 2 PASSED
Test 3 PASSED
Test 4 PASSED
Test 5 PASSED
Test 6 PASSED
Test 7 PASSED
Test 8 PASSED
Test 9 PASSED
Test 10 PASSED
Test 11 PASSED
Test 12 PASSED
Test 13 PASSED
Test 14 PASSED
Test 15 PASSED
Test 16 PASSED
Test 17 PASSED
Test 18 PASSED
Test 19 PASSED
Test 20 PASSED
Test 21 PASSED
Test 22 PASSED
Test 23 PASSED
Test 24 PASSED
Test 25 PASSED
Test 26 PASSED
Test 27 PASSED
Test 28 PASSED
Test 29 PASSED
Test 30 PASSED
Test 31 PASSED
Test 32 PASSED
Test 33 PASSED
Test 34 PASSED
Test 35 PASSED
Test 36 PASSED
Test 37 PASSED
Test 38 PASSED
Test 39 PASSED
Test 40 PASSED
Test 41 PASSED
Test 42 PASSED
Test 43 PASSED
Test 44 PASSED
Test 45 PASSED
Test 46 PASSED
Test 47 PASSED
Test 48 PASSED
Test 49 PASSED
Test 50 PASSED
Test 51 PASSED
Test 52 PASSED
Test 53 PASSED
Test 54 PASSED
Test 55 PASSED
Test 56 PASSED
Test 57 PASSED
Test 58 PASSED
Test 59 PASSED
Test 60 PASSED
Test 61 PASSED
Test 62 PASSED
Test 63 PASSED
Test 64 PASSED
Test 65 PASSED
Test 66 PASSED
Test 67 PASSED
Test